internal/controller/biossettings_controller.go

```diff
@@ -0,0 +1,776 @@
+		log.V(1).Info("object is being deleted")
+		return r.reconcileDeletion(ctx, log, biosSettings)
+	}
+
+	return r.reconcile(ctx, log, biosSettings)
+}
+
+func (r *BiosSettingsReconciler) reconcileDeletion(
+	ctx context.Context,
+	log logr.Logger,
+	biosSettings *metalv1alpha1.BiosSettings,
+) (ctrl.Result, error) {
+	if !controllerutil.ContainsFinalizer(biosSettings, biosSettingsFinalizer) {
+		return ctrl.Result{}, nil
+	}
+	if err := r.cleanupReferences(ctx, log, biosSettings); err != nil {
+		log.Error(err, "failed to cleanup references")
+		return ctrl.Result{}, err
+	}
+	log.V(1).Info("ensured references were cleaned up")
+
+	log.V(1).Info("Ensuring that the finalizer is removed")
+	if modified, err := clientutils.PatchEnsureNoFinalizer(ctx, r.Client, biosSettings, biosSettingsFinalizer); err != nil || modified {
+		return ctrl.Result{}, err
+	}
+
+	log.V(1).Info("biosSettings is deleted")
+	return ctrl.Result{}, nil
+}
+
+func (r *BiosSettingsReconciler) cleanupServerMaintenanceReferences(
+	ctx context.Context,
+	log logr.Logger,
+	biosSettings *metalv1alpha1.BiosSettings,
+) error {
+	if biosSettings.Spec.ServerMaintenanceRef == nil {
+		return nil
+	}
+	// try to get the serverMaintaince created
+	serverMaintenance, err := r.getReferredServerMaintenance(ctx, log, biosSettings.Spec.ServerMaintenanceRef)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return fmt.Errorf("failed to get referred serverMaintenance obj from biosSettings: %w", err)
+	}
+
+	// if we got the server ref. by and its not being deleted
+	if err == nil && serverMaintenance.DeletionTimestamp.IsZero() {
+		// if the serverMaintenace is not created by biosSettings Controller, dont delete.
+		labelsOnMaintenance := serverMaintenance.GetLabels()
+
+		// created by the controller
+		if key, ok := labelsOnMaintenance[biosSettingsCreatorLabel]; ok && key == biosSettings.Name {
+			// if the biosSettings is not being deleted, update the
+			log.V(1).Info("Deleting server maintenance", "serverMaintenance Name", serverMaintenance.Name, "state", serverMaintenance.Status.State)
+			if err := r.Delete(ctx, serverMaintenance); err != nil {
+				return err
+			}
+		} else { // not created by controller
+			log.V(1).Info("server maintenance status not updated as its provided by user", "serverMaintenance Name", serverMaintenance.Name, "state", serverMaintenance.Status.State)
+		}
+	}
+	// if already deleted or we deleted it or its created by user, remove reference
+	if apierrors.IsNotFound(err) || err == nil {
+		err = r.patchMaintenanceRequestRefOnBiosSettings(ctx, log, biosSettings, nil)
+		if err != nil {
+			return fmt.Errorf("failed to clean up serverMaintenance ref in biosSettings status: %w", err)
+		}
+	}
+	return nil
+}
+
+func (r *BiosSettingsReconciler) cleanupReferences(
+	ctx context.Context,
+	log logr.Logger,
+	biosSettings *metalv1alpha1.BiosSettings,
+) (err error) {
+	if err := r.cleanupServerMaintenanceReferences(ctx, log, biosSettings); err != nil {
```

Comment:
Since we are setting ```
return controllerutil.SetControllerReference(biosSettings, serverMaintenance, r.Client.Scheme())
``` 
the `ServerMaintenance` object will be remove by kube once the `BIOSSetting` resource is also removed.